crates/turbopack-node/src/pool.rs

```diff
@@ -560,6 +511,167 @@ impl RunningNodeJsPoolProcess {
+    }
+
+    fn add_queued_task(&mut self) {
+        self.queued_tasks += 1;
+    }
+
+    fn add_cold_process_time(&mut self, time: Duration) {
+        self.total_cold_process_time += time;
+        self.cold_process_count += 1;
+        self.queued_tasks -= 1;
+    }
+
+    fn add_warm_process_time(&mut self, time: Duration) {
+        self.total_warm_process_time += time;
+        self.warm_process_count += 1;
+        self.queued_tasks -= 1;
+    }
+
+    fn estimated_bootup_time(&self) -> Duration {
+        if self.bootup_count == 0 {
+            Duration::from_millis(200)
+        } else {
+            self.total_bootup_time / self.bootup_count
+        }
+    }
+
+    fn estimated_warm_process_time(&self) -> Duration {
+        if self.warm_process_count == 0 {
+            self.estimated_cold_process_time()
+        } else {
+            self.total_warm_process_time / self.warm_process_count
+        }
+    }
+
+    fn estimated_cold_process_time(&self) -> Duration {
+        if self.cold_process_count == 0 {
+            // We assume cold processing is half of bootup time
+            self.estimated_bootup_time() / 2
+        } else {
+            self.total_cold_process_time / self.cold_process_count
+        }
+    }
+    fn wait_time_before_bootup(&self) -> Duration {
+        if self.workers == 0 {
+            return Duration::ZERO;
+        }
+        let booting_workers = self.booting_workers;
+        let workers = self.workers;
+        let warm_process_time = self.estimated_warm_process_time();
+        let expected_completion = self.expected_completion(workers, booting_workers);
+
+        let new_process_duration =
+            self.estimated_bootup_time() + self.estimated_cold_process_time();
+        if expected_completion + warm_process_time < new_process_duration {
+            // Running the task with the existing warm pool is faster
+            return (expected_completion + warm_process_time + new_process_duration) / 2;
+        }
+
+        let expected_completion_with_additional_worker = max(
+            new_process_duration,
+            self.expected_completion(workers + 1, booting_workers + 1),
+        );
+        if expected_completion > expected_completion_with_additional_worker {
+            // Scaling up the pool would help to complete work faster
+            return Duration::ZERO;
+        }
+
+        // It's expected to be faster if we queue the task
+        (expected_completion + expected_completion_with_additional_worker) / 2
+    }
+
+    fn expected_completion(&self, workers: u32, booting_workers: u32) -> Duration {
+        if workers == 0 {
+            return Duration::MAX;
+        }
+        let bootup_time = self.estimated_bootup_time();
+        let cold_process_time = self.estimated_cold_process_time();
+        let warm_process_time = self.estimated_warm_process_time();
+        let expected_full_workers_in = booting_workers * (bootup_time / 2 + cold_process_time);
```

Comment:
Why is it divided by 2?

Reply:
We only know the worker is booting up and not how much time it already spend booting up. It might be shortly before end or just started. If we assume a linear distribution from 0 to `bootup_time` the statistical "expected" time is `bootup_time / 2`